babel/numbers.py

```diff
@@ -20,46 +20,104 @@
 #  - http://www.unicode.org/reports/tr35/ (Appendix G.6)
 import re
 from datetime import date as date_, datetime as datetime_
+from itertools import chain
 
-from babel.core import default_locale, Locale, get_global
+from babel.core import default_locale, Locale, get_global, UnknownLocaleError
 from babel._compat import decimal
+from babel.localedata import locale_identifiers
 
 
 LC_NUMERIC = default_locale('LC_NUMERIC')
 
 
-def get_currency_name(currency, count=None, locale=LC_NUMERIC):
+def list_currency(locale=None):
+    """ Return a list of normalized currency codes.
+
+    .. versionadded:: 2.5.0
+
+    :param locale: filters returned currency codes by the provided locale.
+                   Expected to be a locale instance or code. If no locale is
+                   provided, returns the list of all currencies from all
+                   locales.
+    """
+    # Get locale-scopped currencies.
+    if locale:
+        return Locale.parse(locale).currencies.keys()
+    # Aggregate currency codes from all locales.
+    return set(chain.from_iterable([
+        Locale.parse(l).currencies.keys() for l in locale_identifiers()]))
+
+
+def is_currency(currency_code, locale=None):
+    """ Check the currency code is supported by Babel.
+
+    Accepts a ``locale`` parameter for fined-grained validation, working as
+    the one defined above in ``list_currency()`` method.
+    """
+    try:
+        if currency_code not in list_currency(locale):
+            return False
+    except (ValueError, UnknownLocaleError):
+        return False
+    return True
+
+
+def normalize_currency(currency_code, locale=None):
+    """ Normalize any currency code string.
+
+    Accepts a ``locale`` parameter for fined-grained validation, working as
+    the one defined above in ``list_currency()`` method.
+    """
+    currency_code = currency_code.upper()
+    if not is_currency(currency_code, locale):
+        raise ValueError("Invalid currency code {!r}.".format(currency_code))
+    return currency_code
+
+
+def get_currency_name(currency_code, count=None, locale=LC_NUMERIC):
     """Return the name used by the locale for the specified currency.
 
     >>> get_currency_name('USD', locale='en_US')
     u'US Dollar'
 
     .. versionadded:: 0.9.4
 
-    :param currency: the currency code
+    :param currency: the currency ISO code.
```

Comment:
Oh yes, sorry for that. It was too ambitious and `currency` as it is now is good enough.

I reverted in commit 5ba7c7b0debb650a0f0e80e3ab8a196ac9549193.